### PR TITLE
Animate the gamified dialog

### PR DESCRIPTION
**Pull request by balderdash**
recording 1

The progress earned from each checkbox is a little underwhelming (each step moves the bar maybe 2-3%). I'm wondering if there's some way to highlight it such that it looks more impressive. Also, some of the animation code is a bit hacked up, e.g. implementing a delay by adding an extra StaggeredMotion step.

### CODE REVIEW

File: apps/src/templates/AchievementDialog.jsx [hunk 1]
```diff
@@ -195,26 +205,69 @@ const AchievementDialog = Radium(React.createClass({
...
-                style={{
-                  ...styles.progressForeground,
-                  width: this.props.stageProgress * MAX_PROGRESS_WIDTH
-                }}
-              />
-            </div>
-          </div>
-        }
+        <StaggeredMotion
+          defaultStyles={Array(4).fill({ progress: 0 })}
```
Comment (joshlory) [inline apps/src/templates/AchievementDialog.jsx:210]: Looks like `Array.prototype.fill` isn't supported in IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill#Browser_compatibility

We should be able to use the lodash `_.fill` replacement.

Comment (joshlory) [reply apps/src/templates/AchievementDialog.jsx:210]: Or does our build process polyfill this for us?  I can't remember..

File: apps/src/feedback.js [hunk 2]
```diff
@@ -401,20 +414,29 @@ FeedbackUtils.prototype.calculateStageProgress = function (hintsUsed) {
...
     0.5 * numOneHint / levels.length;
-
-  return 0.3 * passedScore +
+  const oldStageProgress = 0.3 * passedScore +
     0.4 * perfectScore +
     0.3 * hintScore;
+
+  const newPassedLevels = 1;
+  const newPerfectLevels = isPerfect ? 1 : 0;
+  const newHintUsageLevels = hintsUsed === 0 ? 1 : hintsUsed === 1 ? 0.5 : 0;
```
Comment (joshlory) [inline apps/src/feedback.js]: Nit: not a huge fan of nested ternary operators.  If this is the cleanest way to do it let's add parens around the inner ternary.

File: apps/src/templates/AchievementDialog.jsx [hunk 3]
```diff
@@ -195,26 +205,69 @@ const AchievementDialog = Radium(React.createClass({
...
+          styles={prevInterpolatedStyles => prevInterpolatedStyles.map((_, i) => {
+            return i === 0 ?
+              { progress: spring(1, {stiffness: 50, damping: 25}) } :
+              { progress: spring(
+                  prevInterpolatedStyles[i - 1].progress > 0.8 ? 1 : 0,
```
Comment (joshlory) [inline apps/src/templates/AchievementDialog.jsx]: Worth pulling out magic number `0.8` as `const animationOverlap` or similar?

File: apps/src/templates/AchievementDialog.jsx [hunk 4]
```diff
@@ -195,26 +205,69 @@ const AchievementDialog = Radium(React.createClass({
...
-                }}
-              />
-            </div>
-          </div>
-        }
+        <StaggeredMotion
+          defaultStyles={Array(4).fill({ progress: 0 })}
+          styles={prevInterpolatedStyles => prevInterpolatedStyles.map((_, i) => {
+            return i === 0 ?
+              { progress: spring(1, {stiffness: 50, damping: 25}) } :
```
Comment (joshlory) [inline apps/src/templates/AchievementDialog.jsx]: Super tiny nit: inconsistent spacing inside `{ ... }` (space here, no space for e.g. `{fontSize: 24, left: 2, top: 10}`).

File: apps/src/feedback.js [hunk 5]
```diff
@@ -387,6 +397,9 @@ FeedbackUtils.prototype.calculateStageProgress = function (hintsUsed) {
     numZeroHints = 0,
     numOneHint = 0;
   for (let i = 0; i < levels.length; i++) {
+    if (levels[i].ids.indexOf(currentLevelId) !== -1 ) {
+      continue;
+    }
```
Comment (joshlory) [inline apps/src/feedback.js:402]: I wonder if this is a sign that we should be exposing a cleaner interface on `ClientState`.  Are there other places that would want to get the list of passed or perfected dots, collapsing level variants into a single best result?

Feel free to ignore, this shouldn't block check-in for this experiment.

File: apps/src/templates/AchievementDialog.jsx [hunk 6]
```diff
@@ -195,26 +205,69 @@ const AchievementDialog = Radium(React.createClass({
...
+        >
+          {interpolatingValues => {
+              const interpolatingStyles =
+                interpolatingValues.map(val => ({ opacity: val.progress }));
+              return (<div>
+                <div style={styles.checkmarks}>
+                  {this.achievementRow(
+                      true,
+                      locale.puzzleCompleted(),
+                      interpolatingStyles[1])}
```
Comment (joshlory) [inline apps/src/templates/AchievementDialog.jsx:228]: Is the 0 index skipped to get the initial delay when the dialog first appears?

File: apps/src/templates/AchievementDialog.jsx (hunk 1 above)
Comment (balderdash) [reply apps/src/templates/AchievementDialog.jsx:210]: Pretty sure babel-polyfill is taking care of it for us. Either way, I tested on saucelabs and it works fine in IE11.

File: apps/src/templates/AchievementDialog.jsx (hunk 4 above)
Comment (balderdash) [reply apps/src/templates/AchievementDialog.jsx]: Fixed.

File: apps/src/templates/AchievementDialog.jsx (hunk 3 above)
Comment (balderdash) [reply apps/src/templates/AchievementDialog.jsx]: Changed it to `(1 - overlap)`, with an overlap constant of 0.2.

File: apps/src/templates/AchievementDialog.jsx (hunk 6 above)
Comment (balderdash) [reply apps/src/templates/AchievementDialog.jsx:228]: Yeah. Not great, but IMO better than using something like a setTimeout to change state.

File: apps/src/feedback.js (hunk 5 above)
Comment (balderdash) [reply apps/src/feedback.js:402]: Yes, we should. I'm going to ignore you because it's Friday afternoon though.

File: apps/src/feedback.js (hunk 2 above)
Comment (balderdash) [reply apps/src/feedback.js]: Just used a let with a conditional.

**Pull request merged**